pilot/pkg/bootstrap/server.go

```diff
@@ -291,24 +280,30 @@ func NewServer(args *PilotArgs) (*Server, error) {
 	// because it depends on the following conditions:
 	// 1) CA certificate has been created.
 	// 2) grpc server has been generated.
-	s.addStartFunc(func(stop <-chan struct{}) error {
-		if s.ca != nil {
-			s.RunCA(s.secureGRPCServerDNS, s.ca, caOpts, args.Config.ControllerOptions, stop)
-		}
-		return nil
-	})
-
-	if s.leaderElection != nil && enableElection.Get() {
+	if s.ca != nil {
 		s.addStartFunc(func(stop <-chan struct{}) error {
-			// We mark this as a required termination as an optimization. Without this, when we exit the lock is
-			// still held for some time (30-60s or so). If we allow time for a graceful exit, then we can immediately drop the lock.
-			s.requiredTerminations.Add(1)
-			go func() {
-				s.leaderElection.Run(stop)
-				s.requiredTerminations.Done()
-			}()
+			s.RunCA(s.secureGRPCServerDNS, s.ca, caOpts)
 			return nil
 		})
+
+		if s.kubeClient != nil {
+			fetchData := func() map[string]string {
+				return map[string]string{
+					constants.CACertNamespaceConfigMapDataName: string(s.ca.GetCAKeyCertBundle().GetRootCertPem()),
+				}
+			}
+			s.addTerminatingStartFunc(func(stop <-chan struct{}) error {
+				leaderelection.
+					NewLeaderElection(args.Namespace, args.PodName, leaderelection.NamespaceController, s.kubeClient).
+					AddRunFunction(func(stop <-chan struct{}) {
+						log.Infof("Starting namespace controller")
+						nc := NewNamespaceController(fetchData, args.Config.ControllerOptions, s.kubeClient)
```

Comment:
why move namespace controller here

Reply:
before it's nesting 2 addStartFunc calls which ends up behaving weird so I broke it out